_posts/blog/2016-09-06-model-view-viewmodel.md

```diff
@@ -0,0 +1,181 @@
+
+At a first glance, MVVM seems very similar to the Model-View-Presenter pattern, because both of them do a great job in abstracting the view's state and behavior. The Presentation Model abstracts a view independent from a specific user-interface platform whereas the MVVM pattern was created by Microsoft to simplify the **event driven** programming of user interfaces.
+
+If, in the MVP pattern, the Presenter was telling the View directly what to display, in MVVM, **ViewModel exposes streams of events** to which the Views can bind to. Also, the Views notify the ViewModel about different actions. Thus, the MVVM pattern supports two-way data binding between the View and ViewModel and there is a many-to-one relationship between View and ViewModel. View has a reference to ViewModel but **ViewModel has no information about the View**.
+
+<center>
+<picture>
+	<a href="/images/blog/model_view_viewmodel/mvvm.png"><img src="/images/blog/model_view_viewmodel/mvvm.png" alt="Model-View-ViewModel Architecture"></a>
+	<figcaption>Model-View-ViewModel class structure</figcaption>
+</picture>
+</center>
+
+## Model-View-ViewModel at upday
+
+A quick look at the Android posts on the upday blog will instantly reveal what our favorite library is: RxJava. So it’s no wonder that RxJava is the backbone of upday's code. The event driven part required by MVVM is done using RxJava’s `Observable`s. Here's how we apply MVVM at upday, with the help of RxJava:
+
+### DataModel
+
+The DataModel exposes data easily consumable through event streams - RxJava's `Observable`s. It composes data from multiple sources, like the network layer, database or shared preferences and exposes them to whoever needs it. The DataModels hold the entire business logic.
+
+Our strong emphasis on the single responsibility principle leads to creating a DataModel for every feature in the app. For example, we have an ArticleDataModel that contains references to the API service and to the database layer. This DataModel handles the business logic ensuring that the latest news from the database is retrieved, by applying an age filter.
+
+### ViewModel
+
+The ViewModel is a model for the View of the app, an abstraction of the View. The ViewModel retrieves from the DataModel the necessary data, applies the UI logic and then exposes relevant data for the View to consume, together with behaviors for the View. Similar to the DataModel, the ViewModel exposes the data via `Observable`s.
+
+We learned two things about the ViewModel the hard way:
+
+* The ViewModel should expose states for the View, rather than just events. For example, if we need to display the name and the email address of a `User`, rather than creating two streams for this, we create a `DisplayableUser` object that encapsulates the two fields. The stream will emit every time the display name or the email change. Like this, we ensure that our View always displays the current state of the `User`.
```

Comment:
Early you said that MVVM was about events, but here you say states rather than just events. I understand and agree, but it might seem a little contradictory to a reader.